### PR TITLE
Add support for django 1.11

### PR DESCRIPTION
**Pull request by jlward**

### CODE REVIEW

File: form_utils/fields.py [hunk 1]
```diff
@@ -41,9 +41,8 @@ class ClearableFileField(forms.MultiValueField):
     default_file_field_class = forms.FileField
     widget = ClearableFileInput
 
-    def __init__(self, file_field=None, template=None, *args, **kwargs):
```
Comment (jlward) [inline form_utils/fields.py:44]: Cannot pass along `file_field` any longer. We aren't using it and it only complicates things in Django 1.11

File: form_utils/templatetags/form_utils.py [hunk 2]
```diff
@@ -40,7 +40,7 @@ def render(form, template_name=None):
         default = ','.join(['form_utils/better_form.html', default])
     tpl = select_template_from_string(template_name or default)
 
-    return tpl.render(template.Context({'form': form}))
```
Comment (jlward) [inline form_utils/templatetags/form_utils.py:43]: Long story short, this was the main fix.

File: form_utils/widgets.py [hunk 3]
```diff
@@ -59,15 +61,65 @@ def render(self, name, value, attrs=None):
         return mark_safe(output)
 
 
+class ClearableFileInputRenderer(object):
```
Comment (jlward) [inline form_utils/widgets.py:64]: This is hijacking the way django 1.11 renders widgets and making it work with this lib. This code only runs in Django 1.11

File: tests/tests.py [hunk 4]
```diff
@@ -197,7 +197,7 @@ class BetterFormTests(TestCase):
                                 'classes': '',
                                 }),
                     (['reference'],
-                    {
+                     {
```
Comment (jlward) [inline tests/tests.py:200]: A lot of this is flake8

File: tests/tests.py [hunk 5]
```diff
@@ -466,21 +470,40 @@ class BoringForm(forms.Form):
...
-        u'<input type="%(type)s" name="boredom" id="id_boredom" />'
-        u'</li>'
-        u'<li>'
-        u'<label for="id_excitement">Excitement%(suffix)s</label>'
-        u'<input type="%(type)s" name="excitement" id="id_excitement" />'
-        u'</li>'
-        u'</ul>'
-        u'</fieldset>'
-        ) % {'type': number_field_type, 'suffix': label_suffix}
+    @property
```
Comment (jlward) [inline tests/tests.py:475]: `required` change how it works in Django 1.11. These to field converted to properties makes it work in both django versions.

File: tests/tests.py [hunk 6]
```diff
@@ -576,34 +635,16 @@ def test_render(self):
             '<input type="checkbox" name="fieldname_1" />'
             )
 
-    def test_custom_file_widget(self):
```
Comment (jlward) [inline tests/tests.py:579]: We don't care about this functionality. So it's been removed.

File: tests/tests.py [hunk 7]
```diff
@@ -635,11 +676,20 @@ def test_bound_redisplay(self):
         class TestForm(forms.Form):
             f = ClearableFileField()
         form = TestForm(files={'f_0': self.upload})
+        if django.VERSION < (1, 10, 0):
```
Comment (jlward) [inline tests/tests.py:679]: More `required` shenanigans

File: tests/tests.py [hunk 8]
```diff
@@ -680,37 +730,6 @@ def test_cleared_but_file_given(self):
         result = field.clean([self.upload, '1'])
         self.assertEqual(result, self.upload)
 
-    def test_custom_file_field(self):
```
Comment (jlward) [inline tests/tests.py:683]: We don't care about any of this either.

File: tox.ini [hunk 9]
```diff
@@ -1,20 +1,12 @@
 [tox]
 envlist=
-  py26-django{14,15,16},
-  py27-django{14,15,16,17,18,_trunk},
-     py33-django{15,16,17,18,_trunk},
-     py34-django{15,16,17,18,_trunk}
+  py27-django{18,111},
```
Comment (jlward) [inline tox.ini:3]: We only care about python 2.7 and django 1.8/1.11. Ignore everything else.

File: tox.ini [hunk 10]
```diff
@@ -1,20 +1,12 @@
...
 commands=python setup.py test
 
 deps =
-  django14: Django>=1.4, <1.5
-  django15: Django>=1.5, <1.6
-  django16: Django>=1.6, <1.7
-  django17: Django>=1.7, <1.8
   django18: Django>=1.8, <1.9
-  django_trunk: https://github.com/django/django/tarball/master
-  South==0.8.4
```
Comment (jlward) [inline tox.ini:18]: So long sweet prince.

**Review by petereilly-rldatix — APPROVED** [review]
Approved, but I'd like some eyes on it from someone who didn't work on it :stuck_out_tongue:

File: form_utils/widgets.py (hunk 3 above)
Comment (kylegibson-rldatix) [reply form_utils/widgets.py:64]: Can you elaborate more on what ensures this code only runs in django 1.11?

Comment (jlward) [reply form_utils/widgets.py:64]: https://github.com/PolicyStat/django-form-utils/pull/1/files#diff-5928fad501fed3b57076c4e760027bd3R137 This line of code is passing along an instance of this to the render method. It only gets passed along in Django 1.11 (the render method in Djagno 1.8 crashes if passed this instance.

Comment (kylegibson-rldatix) [reply form_utils/widgets.py:64]: `django.VERSION > (1, 10, 0)` would allow django 1.10.x, with x > 0, right?

File: form_utils/widgets.py [hunk 11]
```diff
@@ -59,15 +61,65 @@ def render(self, name, value, attrs=None):
...
+        self.template_string = template_string
+
+    def set_value(self, value):
+        self.input_value = value
+
+    def render(self, template_name, context, request=None):
+
+        input_context = dict(widget={})
+        input_context['widget'].update(context['widget']['subwidgets'][0])
+        # input_context['widget']['value'] = context['widget']['value']
```
Comment (kylegibson-rldatix) [inline form_utils/widgets.py]: Can we remove this commented code here, and below?

File: tests/runtests.py [hunk 12]
```diff
@@ -18,7 +19,16 @@
         MEDIA_URL='/media/',
         STATIC_URL='/static/',
         MIDDLEWARE_CLASSES=[],
-        )
+        TEMPLATES=[
+            {
+                'BACKEND': 'django.template.backends.django.DjangoTemplates',
+                'DIRS': [
+                    '{base}/forms/templates/'.format(base=__path__[0]),
```
Comment (kylegibson-rldatix) [inline tests/runtests.py]: `os.path.join(__path__[0], 'forms', 'templates')`

File: form_utils/widgets.py [hunk 13]
```diff
@@ -77,19 +129,33 @@ def render(self, name, value, attrs=None):
             self.value = value[0]
         else:
             self.value = value
-        return super(ClearableFileInput, self).render(name, value, attrs)
+
+        self.renderer.set_value(value)
+
+        render_kwargs = dict()
+        if django.VERSION > (1, 10, 0):
+            render_kwargs['renderer'] = self.renderer
```
Comment (kylegibson-rldatix) [inline form_utils/widgets.py]: Alternatively, we could do this:

```python
render_kwargs['renderer'] = renderer
if django.VERSION < (1.11.0):
   del render_kwargs['renderer']
```

This way, the latest behavior is always the default, and when support is dropped for older versions, removing that code is easier.

File: form_utils/widgets.py (hunk 3 above)
Comment (jlward) [reply form_utils/widgets.py:64]: Yes, but we are the only people that will ever use this lib and we will only ever be using it with django 1.8 and django 1.11. I specifically chose a number in the middle so I wouldn't have to worry figuring out if `>=` works when comparing tuples.

**Comment by kylegibson-rldatix** [comment]
I wasn't able to do a thorough review of the tests.py file because it wouldn't load in github. I cloned the repo, and did a quick review using git diff.

Other than that, all looks good.

**Pull request merged**